### PR TITLE
Map view roll

### PR DESCRIPTION
**Pull request by schedul-xor**
Added camera roll settings to MapPosition

### CODE REVIEW

File: vtm-web-app/src/org/oscim/web/client/GwtMap.java [hunk 1]
```diff
@@ -175,6 +176,18 @@ public void onValueChange(int val, int max) {
...
             }
         });
+
+        mCameraRollControl = new CameraRollControl("#camera-roll-input");
+        mCameraRollControl.addValueChangeListener(new BuildingSolutionControl.ValueChangeListener() {
+            @Override
+            public void onValueChange(int val, int max) {
+                MapPosition p = mMap.getMapPosition();
+                p.setRoll((float) val / (float) max * 360.0f);
+                mMap.setMapPosition(p);
```
Comment (devemux86) [inline vtm-web-app/src/org/oscim/web/client/GwtMap.java]: Couldn't these lines be replaced with same functionality:
```java
mMap.viewport().setRotation((float) val / (float) max * 360.0f);
mMap.updateMap(true);
```

**Comment by devemux86** [comment]
How is this different than map view rotation (roll suggests different meaning)?
Because as indicated at commit comment, this is already possible via:
```java
map.viewport().setRotation(...);
```

File: vtm-web-app/src/org/oscim/web/client/GwtMap.java (hunk 1 above)
Comment (devemux86) [reply vtm-web-app/src/org/oscim/web/client/GwtMap.java]: Or even:
```java
p.setBearing((float) val / (float) max * 360.0f);
```

**Comment by devemux86** [comment]
And there is already the `MapPosition.setBearing(...)` if want to play with map position.

File: vtm-web/src/org/oscim/gdx/client/MapUrl.java [hunk 2]
```diff
@@ -83,36 +86,39 @@ else if (p.startsWith("tilt="))
...
+        float lon = (float) (MercatorProjection.toLongitude(pos.x) * 1000000);
+        float rot = ((pos.bearing) % 360) * 1000000;
+        float roll = ((pos.roll) % 360) * 1000000;
+        float tilt = pos.tilt * 1000000;
+        float zoom = (float) pos.zoomLevel * 1000000;
 
-        if (curZoom != pos.zoomLevel || curLat != lat || curLon != lon
-                || curTilt != rot || curRot != (int) (pos.bearing)) {
+        if (curZoom != zoom || curLat != lat || curLon != lon
+                || curTilt != rot || curRot != rot || curRoll != roll) {
```
Comment (devemux86) [inline vtm-web/src/org/oscim/gdx/client/MapUrl.java]: BTW why the change to floats since compare is again with fixed precision?

File: vtm/src/org/oscim/map/ViewController.java [hunk 3]
```diff
@@ -247,6 +261,9 @@ private void updateMatrices() {
         mTmpMatrix.setTranslation(0, mPivotY * mHeight, 0);
         mViewMatrix.multiplyLhs(mTmpMatrix);
 
+        mTmpMatrix.setRotation(mPos.roll, 0, 0, 1);
```
Comment (devemux86) [inline vtm/src/org/oscim/map/ViewController.java]: The axis doesn't seem right for actual "roll"?

**Comment by devemux86** [comment]
Current implementation seems similar to existing bearing, i.e. vertical axis (yaw).
If you mean [longitudinal axis (roll)](https://en.wikipedia.org/wiki/Aircraft_principal_axes), then `ViewController` calculations could need improvement (see commit comment).

File: vtm/src/org/oscim/map/Viewport.java [hunk 4]
```diff
@@ -52,6 +52,9 @@
     protected float mMinBearing = -180;
     protected float mMaxBearing = 180;
 
+    protected float mMinRoll = 0;
+    protected float mMaxRoll = 360;
```
Comment (devemux86) [inline vtm/src/org/oscim/map/Viewport.java]: Limits better be -180 to 180 degrees.

**Comment by schedul-xor** [comment]
Thanks for your review.

Yes you're right, I was expecting longitudinal axis (roll). It makes no difference when the map is not tilted, but is needed to make rendered map match to photos/videos taken by tilted surveillance cameras.

I will look in the commit comments.

**Comment by devemux86** [comment]
Probably something like `mTmpMatrix.setRotation(mPos.roll, 0, 1, 0);` is needed.

File: vtm-web/src/org/oscim/gdx/client/MapUrl.java (hunk 2 above)
Comment (schedul-xor) [reply vtm-web/src/org/oscim/gdx/client/MapUrl.java]: Now I understood the reason why the values are compared using integer ... it was to fix the precision. I modified these just because I thought integer degrees was very low precision for bearing, tilt and rolls. But that's not the main issue. I'll roll it back.

**Comment by devemux86** [comment]
Indeed position bearing (yaw) and tilt (pitch) complicate the roll, as the 3 motions should be done around their rotated axes.

A nice way to see them all in action and test is Android emulator extended controls:

controls

**Comment by devemux86** [comment]
@schedul-xor any news or need any help?

**Comment by schedul-xor** [comment]
@devemux86 Not any updates yet, I'm wondering what I should be working from.

 * vtm-web/src/org/oscim/gdx/client/MapUrl.java strange refactoring (at least should not be done at this branch) -> I'll revert it as soon as possible.
 * Roll values to be normalized to -180 .. 180, not 0 .. 360 -> I'll revert it as soon as possible.
 * How about pitches and yaws? -> I believe pitches are similar to tilt, yaws are similar to bearing.
 * Need `mTmpMatrix.setRotation(mPos.roll, 0, 1, 0);` -> Not sure why. I tested my code in vtm-web-app, the second lever located at the left-top corner will roll the view, even when the map is tilted.

**Comment by devemux86** [comment]
> How about pitches and yaws? -> I believe pitches are similar to tilt, yaws are similar to bearing.

The library already supports Yaw (via bearing) and Pitch (via tilt).

> I tested my code in vtm-web-app, the second lever located at the left-top corner will roll the view, even when the map is tilted.

What about when not tilted?
Then it does not roll, it just rotates (like yaw), which is not the expected behavior.
Android emulator extended controls have different implementation, so should decide behavior first.

**Comment by devemux86** [comment]
BTW `mTmpMatrix.setRotation(mPos.roll, 0, 1, 0);` reflects exactly Android emulator extended controls.

File: vtm/src/org/oscim/layers/marker/MarkerRenderer.java [hunk 5]
```diff
@@ -133,6 +133,7 @@ else if (it.x < -flip)
         /* keep position for current state */
         mMapPosition.copy(v.pos);
         mMapPosition.bearing = -mMapPosition.bearing;
+        mMapPosition.roll = -mMapPosition.roll;
```
Comment (devemux86) [inline vtm/src/org/oscim/layers/marker/MarkerRenderer.java]: Not sure if that is needed, since neither tilt is used either.

File: vtm/src/org/oscim/layers/marker/ClusterMarkerRenderer.java [hunk 6]
```diff
@@ -311,6 +311,7 @@ else if (it.x < -flip)
         /* keep position for current state */
         mMapPosition.copy(v.pos);
         mMapPosition.bearing = -mMapPosition.bearing;
+        mMapPosition.roll = -mMapPosition.roll;
```
Comment (devemux86) [inline vtm/src/org/oscim/layers/marker/ClusterMarkerRenderer.java]: Not sure if that is needed, since neither tilt is used either.

**Comment by devemux86** [comment]
I merged it via 49476e17bbd9ae802b36d56c1f31c196340e416b with all aforementioned improvements and roll around y-axis.

I included the fix for actual roll in 05d0b2b116bfd4973709c8d342f2ae62fc53ffca for easy reference.
Map view is not an aircraft, but that should be the expected behavior?

**Comment by schedul-xor** [comment]
Thank you so much @devemux86, the work is perfectly matching to what I've expected.

**Comment by devemux86** [comment]
Tilt (Pitch) 65° and Roll 32°:

roll
